### PR TITLE
Add SORTED_NUMERIC type for NumberFieldType 

### PR DESCRIPTION
**Pull request by wuyunfeng**
Since `MappedFieldType` comment remove this `docvalues` flag and use `docValuesType`,  we should use `docValuesType()` to decide which this Field has docvalue. But when encounter the number field, obtain the `MappedFieldType` and use `docValuesType` method return `DocValuesType.NONE`, I think this is so mistake.  I find all number field in Elasticsearch use `SORTED_NUMERIC` type, so adding it to the final class `NumberFieldType`, does it OK?  When added this, our plugin works well.

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@wuyunfeng  Thank you for your PR.  We indeed need to change for a check to `docValuesType` instead of `hasDocValues`? Would you have time and be interested to do more refactoring along this lines?  For example, to modify `MapperFieldType::hasDocValues()` to return if docValuesType == DocValuesType.NONE?

**Comment by wuyunfeng** [comment]
@mayya-sharipova Thanks for your reply. I would like to try it. But then，maybe some modifications involving lot of refactor?

**Comment by mayya-sharipova** [comment]
@wuyunfeng Right, it may, but from what I can judge it may touch a lot of files, but modifications in each of the file will be small.  We can also break the change into several PRs.

**Comment by wuyunfeng** [comment]
@mayya-sharipova  maybe should have a schedule... ^o^, divided into serveral PRs is a good idea

**Comment by wuyunfeng** [comment]
@mayya-sharipova  I modify some core MapperFieldType first,  can you give your review message?  Adopt your advice  divide into serveral PRs

**Comment by wuyunfeng** [comment]
ping @mayya-sharipova

**Review by mayya-sharipova — CHANGES_REQUESTED** [review]
@wuyunfeng  Thank you for continuing your work on this PR.  Good changes.  
The main comment is that  I think we don't need `doSetDocValueType` in any of the files.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 1]
```diff
@@ -111,9 +111,15 @@ public T store(boolean store) {
...
 
+        public void doSetDocValueType(boolean docValues) {
+
+        }
+
+
         public T docValues(boolean docValues) {
             this.fieldType.setHasDocValues(docValues);
             this.docValuesSet = true;
+            doSetDocValueType(docValues);
```
[the remark this answers is not in the document]
Comment (wuyunfeng) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java]: Good idea, I will put this line into  MapperFieldType setHasDocValues(). BTW,  delete `docValues` flag  is OK

**Comment by wuyunfeng** [comment]
@mayya-sharipova  some XXXAggregatorTests may use the XXXFieldType() constructor directly, and the setDocValue(true),  What should we change for this scenario?

**Comment by mayya-sharipova** [comment]
@wuyunfeng Thanks a lot, this is a good change. And modifying tests is great. I have started the tests to see that we have not broken anything.

**Comment by mayya-sharipova** [comment]
@elasticmachine  test this please

**Comment by wuyunfeng** [comment]
@mayya-sharipova  Thanks for your comment. I only modify a fraction of XXXTest file but not all, I find lots of test files use the XXXFieldType constructor directly then invoke the setHasDocValues(true), in this way not set the docValueType.  Maybe need to modify the XXXAggregatorTest and some other files
BTW,  I want to remove the `docValues` flag from the `MappedFieldType`, what do you think?

**Comment by mayya-sharipova** [comment]
@wuyunfeng  I would suggest let's keep the changes for this PR as minimal as possible. 
For this PR we can only keep changes for:
-  all *FieldMapper.java files
- MappedFieldType.java
- and only *Tests.java files that break (for example keep the changes to `KeywordFieldMapperTests.java`, but don't change `*AggregatorTests.java` files if they don't break.  All  tests `*AggregatorTests.java` should pass, so let's not modify them in this PR).

In the subsequent PR, we can remove `docValues` field from `MappedFieldType`

**Comment by wuyunfeng** [comment]
@mayya-sharipova OK, I will restore *AggregatorTests.java later

**Comment by wuyunfeng** [comment]
@mayya-sharipova  I have rollbacked xxxAggregatorTests, I think KeywordFieldMapperTests should fail without changing

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by wuyunfeng** [comment]
@mayya-sharipova  some UT fails seems not to related with this PR, and somewhere needed to be changed?

**Comment by wuyunfeng** [comment]
@elasticmachine test this

**Comment by mayya-sharipova** [comment]
@elasticmachine test this please

**Comment by wuyunfeng** [comment]
@mayya-sharipova  
1. Some FieldMapper  ` fields.add(new Field(fieldType().name(), value, fieldType()) ` resulting in NullPointerException when index doc,  `noDocValueFieldType()` getting rid of the defalut docValue enable.
2.  Some UnitTest using clone method generate FieldType,  `this.docValues = ref.hasDocValues();` resulting `docValues` inconsistency

**Comment by wuyunfeng** [comment]
@mayya-sharipova

**Comment by mayya-sharipova** [comment]
@wuyunfeng  Sorry for a delay. This turned out to be more complicated than I thought. 
 I will need to discuss this change with my team. I will get back to you next wee

**Comment by wuyunfeng** [comment]
@mayya-sharipova 
 

> Can you please also explain why do you need docValuesType for your plugin?
    
   Our plugin `Doris On Elasticsearch` would prefer retrieve value from  `docvalue` , you can see this issue for Doris（incubator-doris）: [Doris On Elasticsearch](incubator-doris) , Just like elasticsearch-hadoop for spark analysis. In Our first version we write native Luece Query for `query phase`, and use the origin `Fetch Phase`,  `Doris On Elasticsearch` needs to know wheather fields has `docvalue`, if has we would retrieve the value from the `docvalue` through add this field to  `docvalue_fields: [ fieldA, fieldB]` context.  
     Just as you see, when i use `MappedFieldType:docValues` , `Remove this TODO item` caught my attention. If  using this `docValues`, I worried Elasticseach Team will remove this value in future release version. I decided to use `docValuesType` instead of `docValues`, until `docValuesType` always return `DocValuesType.NONE`.

> The problem is that one Elasticsearch MappedFieldType usually corresponds to two Lucene's FieldType - as we create 2 fields: 1 must be with some DocValues and another must be with DocValuesType.NONE.

   Yeah,  use different format value for `index field`, `store field` and `doc_value field`  from same `Field` is a great idea, so I add `noDocValueFieldType ()` method to strip the `docValueType` for normal field.

> We have two potential resolutions: 1. Remove this TODO item, keep the code and MappedFieldType:docValues as is  2. Something suggested in this PR, set in all *FieldMapper in Defaults the corresponding docValuesType FIELD_TYPE.setDocValuesType, but when creating a usual field (not DocValuesField) make sure that it is created with a fieldType where DocValuesType.NONE

   1.  `docValuesType` would still return` DocValuesType.NONE`, maybe mistaking another guys in the future. In this way, we should add note comment for `docValuesType`  when `Remove this TODO item`
  2.  maybe `noDocValueFieldType ` is not a good resolution but can make sure ` DocValuesType.NONE` assign  normal field, maybe the function name need to be changed or the location need to change.

**Comment by wuyunfeng** [comment]
@mayya-sharipova another alternative way, can set the DocValueType when add DocValueField for this field

**Comment by wuyunfeng** [comment]
@mayya-sharipova

**Comment by jpountz** [comment]
The TODO about using docValueType rather than the `docValues` flag made sense at that time when things were simpler than now. However I think a better option now would be to decouple MappedFieldType from Lucene's FieldType. It is a bit of a hack today, but this is almost what we started doing for numeric fields by using `indexOptions` to figure out whether a field is indexed even though numeric fields are no longer using the inverted index, they are using points.

**Comment by wuyunfeng** [comment]
@mayya-sharipova it does not matter. enjoy to discuss with you ^o^

**Pull request closed without merging**